sparse/tpls/KokkosSparse_spmv_tpl_spec_decl.hpp

```diff
@@ -404,76 +487,85 @@ void spmv_rocsparse(const Kokkos::HIP& exec,
-  KOKKOS_IMPL_HIP_SAFE_CALL(hipMalloc(&tmp_buffer, buffer_size));
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_spmv_ex(
-      handle, myRocsparseOperation, &alpha, Aspmat, vecX, &beta, vecY,
-      compute_type, alg, rocsparse_spmv_stage_auto, &buffer_size, tmp_buffer));
+  // Create or update VecX/Y descr
+  if (!spmv_helper->initialized[idx]) {
+    // Create rocsparse dense vectors for X and Y
+    KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_create_dnvec_descr(
+        &spmv_helper->vecX[idx], x.extent_int(0), x_data,
+        rocsparse_compute_type<typename XVector::non_const_value_type>()));
+    KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_create_dnvec_descr(
+        &spmv_helper->vecY[idx], y.extent_int(0), y_data,
+        rocsparse_compute_type<typename YVector::non_const_value_type>()));
+  } else {
+    // Update vecX/Y pointers so that even if x/y have new pointers, we
+    // are still good
+    KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(
+        rocsparse_dnvec_set_values(spmv_helper->vecX[idx], x_data));
+    KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(
+        rocsparse_dnvec_set_values(spmv_helper->vecY[idx], y_data));
+  }
+
+  // Allocate buffer when not initialized or algorithm changed
+  if (!spmv_helper->initialized[idx] || alg != spmv_helper->alg[idx]) {
+    KOKKOS_IMPL_HIP_SAFE_CALL(hipFree(spmv_helper->buffer[idx]));
+#if KOKKOSSPARSE_IMPL_ROCM_VERSION >= 50400
+    KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(
+        rocsparse_spmv_ex(handle, opA, &alpha, spmv_helper->Aspmat,
+                          spmv_helper->vecX[idx], &beta, spmv_helper->vecY[idx],
+                          mat_value_type, alg, rocsparse_spmv_stage_buffer_size,
+                          &spmv_helper->buffer_size[idx], nullptr));
 #else
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(
-      rocsparse_spmv(handle, myRocsparseOperation, &alpha, Aspmat, vecX, &beta,
-                     vecY, compute_type, alg, &buffer_size, tmp_buffer));
-  KOKKOS_IMPL_HIP_SAFE_CALL(hipMalloc(&tmp_buffer, buffer_size));
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(
-      rocsparse_spmv(handle, myRocsparseOperation, &alpha, Aspmat, vecX, &beta,
-                     vecY, compute_type, alg, &buffer_size, tmp_buffer));
+    KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_spmv(
+        handle, opA, &alpha, spmv_helper->Aspmat, spmv_helper->vecX[idx], &beta,
+        spmv_helper->vecY[idx], mat_value_type, alg,
+        &spmv_helper->buffer_size[idx], nullptr));
 #endif
-  KOKKOS_IMPL_HIP_SAFE_CALL(hipFree(tmp_buffer));
 
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_destroy_dnvec_descr(vecY));
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_destroy_dnvec_descr(vecX));
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_destroy_spmat_descr(Aspmat));
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_destroy_mat_descr(Amat));
+    KOKKOS_IMPL_HIP_SAFE_CALL(
+        hipMalloc(&spmv_helper->buffer[idx], spmv_helper->buffer_size[idx]));
+
+#if KOKKOSSPARSE_IMPL_ROCM_VERSION >= 50400
```

Comment:
@lucbv Note that I removed the old ` KOKKOSSPARSE_IMPL_ROCM_VERSION >= 60000` stuff and prefer `rocsparse_spmv_ex` over `rocsparse_spmv`